pyface/util/font_parser.py

```diff
@@ -0,0 +1,141 @@
+
+
+class FontParseError(ValueError):
+    """An exception raised when font parsing fails."""
+    pass
+
+
+def simple_parser(description):
+    """An extremely simple font description parser.
+
+    The parser is simple, and works by splitting the description on whitespace
+    and examining each resulting token for understood terms:
+
+    Size
+        The first numeric term is treated as the font size.
+
+    Weight
+        The following weight terms are accepted: 'thin', 'extra-light',
+        'light', 'regular', 'medium', 'demi-bold', 'bold', 'extra-bold',
+        'heavy', 'extra-heavy'.
+
+    Stretch
+        The following stretch terms are accepted: 'ultra-condensed',
+        'extra-condensed', 'condensed', 'semi-condensed', 'semi-expanded',
+        'expanded', 'extra-expanded', 'ultra-expanded'.
+
+    Style
+        The following style terms are accepted: 'italic', 'oblique'.
+
+    Variant
+        The following variant terms are accepted: 'small-caps'.
+
+    Decorations
+        The following decoration terms are accepted: 'underline',
+        'strikethrough', 'overline'.
+
+    Generic Families
+        The following generic family terms are accepted: 'default', 'fantasy',
+        'decorative', 'serif', 'roman', 'cursive', 'script', 'sans-serif',
+        'swiss', 'monospace', 'modern', 'typewriter', 'teletype'.
+
+    In addtion, the parser ignores the terms 'pt', 'point', 'px', and 'family'.
+    Any remaining terms are combined into the typeface name.  There is no
+    expected order to the terms.
+
+    This parser is roughly compatible with the various ad-hoc parsers in
+    TraitsUI and Kiva, allowing for the slight differences between them and
+    adding support for additional options supported by Pyface fonts, such as
+    stretch and variants.
+
+    Parameters
+    ----------
+    description : str
+        The font description to be parsed.
+
+    Returns
+    -------
+    properties : dict
+        Font properties suitable for use in creating a Pyface Font.
+
+    Notes
+    -----
+    This is not a particularly good parser, as it will fail to properly
+    parse something like "10 pt times new roman" or "14 pt computer modern"
+    since they have generic font names as part of the font face name.
+    """
+    face = []
+    generic_family = ""
+    size = None
+    weight = "normal"
+    stretch = "normal"
+    style = "normal"
+    variants = set()
+    decorations = set()
+    for word in description.lower().split():
+        if word in NOISE:
+            continue
+        elif word in GENERIC_FAMILIES:
+            generic_family = word
+        elif word in WEIGHTS:
+            weight = word
+        elif word in STRETCHES:
+            stretch = word
+        elif word in STYLES:
+            style = word
+        elif word in VARIANTS:
+            variants.add(word)
+        elif word in DECORATIONS:
+            decorations.add(word)
+        else:
+            if size is None:
```

Comment:
It may not matter much, but from reading the code it looks as though there's an inconsistency between the first-one-wins handling for `size` and the last-one-wins handling for `weight`, `stretch`, etc.

Reply:
I agree it's a bit weird.  This parser is deliberately following the approach of the code here:
https://github.com/enthought/traitsui/blob/b7c38c7a47bf6ae7971f9ddab70c8a358647dd25/traitsui/qt4/font_trait.py#L92-L111
and here:
https://github.com/enthought/traitsui/blob/b7c38c7a47bf6ae7971f9ddab70c8a358647dd25/traitsui/wx/font_trait.py#L74-L91

The idea is that it should be possible to switch to this parser with a minimum of behaviour change; and then later we can have a "proper" well-defined parser which people can migrate to in a more controlled way.